ipc-queue/src/fifo.rs

```diff
@@ -0,0 +1,404 @@
+    let inner = FifoInner::from_arc(arc);
+    let tx = Sender { inner: inner.clone(), synchronizer: s.clone() };
+    let rx = Receiver { inner, synchronizer: s };
+    (tx, rx)
+}
+
+pub fn bounded_async<T, S>(len: usize, s: S) -> (AsyncSender<T, S>, AsyncReceiver<T, S>)
+where
+    T: Identified,
+    S: AsyncSynchronizer,
+{
+    let arc = Arc::new(Fifo::new(len));
+    let inner = FifoInner::from_arc(arc);
+    let tx = AsyncSender { inner: inner.clone(), synchronizer: s.clone() };
+    let rx = AsyncReceiver { inner, synchronizer: s };
+    (tx, rx)
+}
+
+struct Fifo<T> {
+    data: Box<[T]>,
+    offsets: Box<AtomicUsize>,
+}
+
+impl<T: Identified> Fifo<T> {
+    fn new(len: usize) -> Self {
+        assert!(
+            len.is_power_of_two(),
+            "Fifo len should be a power of two"
+        );
+        let mut data = Vec::with_capacity(len);
+        data.resize_with(len, T::empty);
+        Self {
+            data: data.into_boxed_slice(),
+            offsets: Box::new(AtomicUsize::new(0)),
+        }
+    }
+}
+
+enum Storage<T> {
+    Shared(Arc<Fifo<T>>),
+    Static,
+}
+
+impl<T> Clone for Storage<T> {
+    fn clone(&self) -> Self {
+        match self {
+            Storage::Shared(arc) => Storage::Shared(arc.clone()),
+            Storage::Static => Storage::Static,
+        }
+    }
+}
+
+pub(crate) struct FifoInner<T> {
+    data: NonNull<UnsafeCell<[T]>>,
+    offsets: NonNull<AtomicUsize>,
+    storage: Storage<T>,
+}
+
+impl<T> Clone for FifoInner<T> {
+    fn clone(&self) -> Self {
+        Self {
+            data: self.data.clone(),
+            offsets: self.offsets.clone(),
+            storage: self.storage.clone(),
+        }
+    }
+}
+
+impl<T: Identified> FifoInner<T> {
+    pub(crate) unsafe fn from_descriptor(descriptor: FifoDescriptor<T>) -> Self {
+        assert!(
+            descriptor.len.is_power_of_two(),
+            "Fifo len should be a power of two"
+        );
+        let data_slice = std::slice::from_raw_parts_mut(descriptor.data, descriptor.len);
+        Self {
+            data: NonNull::new_unchecked(data_slice as *mut [T] as *mut UnsafeCell<[T]>),
+            offsets: NonNull::new_unchecked(descriptor.offsets as *mut AtomicUsize),
+            storage: Storage::Static,
+        }
+    }
+
+    fn from_arc(fifo: Arc<Fifo<T>>) -> Self {
+        unsafe {
+            Self {
+                data: NonNull::new_unchecked(
+                    fifo.data.as_ref() as *const [T] as *mut [T] as *mut UnsafeCell<[T]>
+                ),
+                offsets: NonNull::new_unchecked(
+                    fifo.offsets.as_ref() as *const AtomicUsize as *mut AtomicUsize
+                ),
+                storage: Storage::Shared(fifo),
+            }
+        }
+    }
+
+    /// Consumes `self` and returns a FifoDescriptor. If `self` was created
+    /// using `from_arc`, it leaks the internal `Arc` copy to ensure the
+    /// resulting descriptor is valid for `'static` lifetime.
```

Comment:
This is not great. I think you should return some type that can produce a `FifoDescriptor<T>` but that the caller needs to hold on to for as long as it knows that the other end is using it (e.g. for as long as the enclave lives).